### PR TITLE
CT-2368 closed cases report for offender sar

### PR DESCRIPTION
**Pull request by ymao2**
## Description

There has been a closed report template for all case types including FOI, SAR, etc. but the problem with that template is that there will be some columns meaningful to one of case type but not irrelevant to other case types, it becomes worse for Offender SAR type,  this PR is to 
- create a close report for Offender SAR specific
- have the flexibility to allow adding another close report easily in the future 
- solve the report categories permission issues to show up the reports which is allowed to be seen by the current user  

## Self-review checklist

* [ ] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [X] (3) Tests passing
* [X] (4) Branch ready to be merged (not work in progress)
* [X] (5) No superfluous changes in diff
* [X] (6) No TODO's without new ticket numbers
* [X] (7) PR Prefixed with ticket number e.g. `CT-7654 ...`

### Screenshots

- for other type
Screenshot 2020-07-30 at 17 27 20
Screenshot 2020-07-30 at 17 27 30

- for offender sar
Screenshot 2020-07-30 at 17 27 44
Screenshot 2020-07-30 at 17 27 54

### Related JIRA tickets

https://dsdmoj.atlassian.net/secure/RapidBoard.jspa?rapidView=25&modal=detail&selectedIssue=CT-2368
### Manual testing instructions

1 - login as branson user
2 - choose reports tabl
3 - click "Create report " button
4 - Choose "Offender SAR" and the sub report of "Closed cases report"
5 - choose the time frame
9 - confirm to  "Create custom report"
10 - click the link to download it from flash message

### CODE REVIEW

File: app/models/warehouse/case_report.rb [hunk 1]
```diff
@@ -163,6 +163,17 @@ def dequote_and_truncate(text)
       def humanize_boolean(boolean)
         boolean ? 'Yes' : nil
       end
+
+      private 
+      def get_requester_type(kase)
+        if kase.offender_sar?
+          kase.third_party? ? kase.third_party_relationship : 'data subject'
```
Comment (Polsonby) [inline app/models/warehouse/case_report.rb:170]: Tiny thing, but could we get rid of this check for offender_sar? 

I think if we added a `requester_type` method to the model offender.rb we could put line 51 back how it was

File: app/models/report_type.rb [hunk 2]
```diff
@@ -54,12 +63,19 @@ def self.respond_to_missing?(meth, include_private = false)
...
+    all_reports = {
       foi: self.standard.foi.order(:full_name),
       sar: self.standard.sar.order(:full_name),
       offender_sar: self.standard.offender_sar.order(:full_name)
     }
+    allowed_reports = []
+    self.get_report_categories(correspondence_types).each do | report_category |
+      if !REPORT_TYPE_CORRESPONDENCE_TYPE_MAP[report_category].nil?
+        allowed_reports |= all_reports[REPORT_TYPE_CORRESPONDENCE_TYPE_MAP[report_category]]
+      end
```
Comment (willmcb) [inline app/models/report_type.rb:76]: Is the intention here to ensure that `allowed_reports` has only unique values in it - the bitwise OR (`|=`) with assignment is not that common to see so just checking that its not meant to be `||=`?

**Review by willmcb — APPROVED** [review]
lgtm 👍

File: app/services/stats/etl/closed_cases.rb [hunk 3]
```diff
@@ -2,58 +2,14 @@
...
-        #
-        # Ordering of generation matters, each function call creates a temp CSV
-        @_temp_files = generate_header_fragment + generate_data_fragments
-        self
-      end
 
-      def transform
-        self
+      def result_name
+        RESULTS_NAME
```
Comment (Polsonby) [inline app/services/stats/etl/closed_cases.rb:10]: Super tiny thing, but could you clarify what this method's for?

The fact that it has both plural and singlular forms of "result" here caught my eye

Comment (ymao2) [reply app/services/stats/etl/closed_cases.rb:10]: the method is return the file name of csv zip file (used in func: transform()),  the original code uses the "result_name",  I changed the constant name to keep consistent

**Review by Polsonby — APPROVED** [review]
Great stuff :)

**Pull request merged**